test/Test/MockApi.hs

```diff
@@ -0,0 +1,227 @@
+--
+-- This system resolves around setting expectations.  An expectation represents a
+-- request we expect to be made and the result of that request.  Additionally, the
+-- expectation may be that the request is made only once or many times.
+--
+-- The request must be a @FossaApiClientF@ and must exactly be the request
+-- expected.  This is due to some limitations on matching types at run time.
+-- Expectations are matched in the order they are set up.  An assertion failure
+-- will be raised if a request is made that doesn't have a matching expectation.
+--
+-- The result can either be a return value or to die with a fatal diagnostic
+-- error.
+--
+-- Set up expectations using the `returnsOnce`, `alwaysReturns` and `fails`
+-- helpers.
+--
+-- @
+--     -- Set up basic expectation
+--     GetOrganization `alwaysReturns` Fixtures.organization
+--     GetProject revision `returnsOnce` Fixtures.project
+--
+--     -- Expect get issues to be called and fail, then return successfully when
+--     -- called again.
+--     GetIssues revision `fails` "Mock API error"
+--     GetIssues revision `returnsOnce` Fixtures.issues
+-- @
+--
+-- An expectation is considered satisfied if it had a limited number of invocations
+-- and it has been matched that many times.  If any expectations are unsatisfied at
+-- the end of the test, an test assertion failure will be raised.
+module Test.MockApi (
+  ApiExpectation,
+  FossaApiClientMockC,
+  MockApi (..),
+  MockApiC (runMockApiC),
+  alwaysReturns,
+  assertAllSatisfied,
+  fails,
+  returnsOnce,
+  runMockApi,
+  runApiWithMock,
+) where
+
+import Control.Algebra (Algebra (..), Has, send, type (:+:) (..))
+import Control.Carrier.Simple (SimpleC, interpret)
+import Control.Carrier.State.Strict (StateC (StateC), runState)
+import Control.Effect.Diagnostics (Diagnostics, fatalText)
+import Control.Effect.FossaApiClient (FossaApiClientF (..))
+import Control.Effect.Lift (Lift, sendIO)
+import Control.Effect.State (State, get, modify, put)
+import Control.Monad (guard)
+import Control.Monad.Trans (MonadIO)
+import Data.Kind (Type)
+import Data.List (intercalate)
+import Data.Text (Text)
+import Test.HUnit (assertFailure)
+
+data MockApi (m :: Type -> Type) a where
+  MockApiOnce :: FossaApiClientF a -> a -> MockApi m ()
+  MockApiAlways :: FossaApiClientF a -> a -> MockApi m ()
+  MockApiFails :: FossaApiClientF a -> Text -> MockApi m ()
+  MockApiRunExpectations :: FossaApiClientF a -> MockApi m (Maybe (ApiResult a))
+  AssertUnexpectedCall :: FossaApiClientF a -> MockApi m a
+  AssertAllSatisfied :: MockApi m ()
+
+-- | Specifies how to handle repetition of a request
+data ExpectationRepetition = Once | Always deriving (Eq, Ord, Show)
+
+-- | The result of a call can be either a value or a diagnostic failure
+data ApiResult a = Return a | Die Text deriving (Eq, Ord, Show)
+
+-- | An expectation of an API call made up of the request and response.
+data ApiExpectation where
```

Comment:
Can we write a short document explaining the relationship between these mocking structures?

The doc I think we need is this: if you had only laid out the types and the document, someone else could have independently written all of the utilities and the effect handler basically the same as you've done here.

Reply:
Also, I'd like to understand how the mocking works by only reading docs and comments.  We have a precedent for documenting complex solutions like this [in our contributing docs](https://github.com/fossas/fossa-cli/tree/master/docs/contributing).

Reply:
There's no need to duplicate the info, the missing info is my bigger concern.

Reply:
Got it, I can be more specific.  I bet I was having the inverse problem trying to express what was missing.

Reply:
Few things I'd like to see added:

- How does `handleRequest` work, in the abstract?
- How are we hijacking the `FossaApiClient` effect?

The ideal audience for effects shenanigans is almost always "haskell dev that can use effects in basic forms, but may not understand how to write them or how they work in depth, and doesn't necessarily need to learn yet".  So, for the effect rewriting, it's worth pointing out that effects only have meaning because of their handlers, and `runApiWithMock` is using a completely different handler that _requires_ a `MockApi` effect, stuff like that.  I only understand that stuff because I've been diving into the code and have experience with the effect system, but we can make it easier for others to understand this, even if the code is the only way to get ALL of the details.

Reply:
Okay, I think maybe I got that all in a doc.